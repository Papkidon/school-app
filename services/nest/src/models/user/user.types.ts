import { z } from 'zod';

export const userCreateSchema = z
  .object({
    data: z.object({
      username: z.string().min(2).max(20),
      password: z.string().min(8).max(20),
    }),
  })
  .strict();

export const userUpdateSchema = userCreateSchema.deepPartial();

export type userType = {
  data: {
    username: string;
    password: string;
  };
};
