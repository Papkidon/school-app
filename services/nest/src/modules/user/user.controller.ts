import {
  Body,
  Controller,
  Delete,
  Get,
  HttpException,
  HttpStatus,
  Param,
  Patch,
  Post,
} from '@nestjs/common';
import {
  userCreateSchema,
  userUpdateSchema,
} from '../../models/user/user.types';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { User } from '@prisma/client';
import { userType } from '../../models/user/user.types';
import { DeepPartial, idSchema } from '../../models/utility.types';
import { ZodValidationPipe } from '../../pipes/zod.validation.pipe';
import { UserService } from './user.service';

@Controller('api/v1/user')
export class UserController {
  constructor(private readonly userService: UserService) {}

  @Get('')
  async findAll(): Promise<User[]> {
    try {
      const result = await this.userService.findAll();

      if (!Object.keys(result).length) {
        throw new HttpException(
          'No user data exists in the database',
          HttpStatus.NOT_FOUND,
        );
      }

      return result;
    } catch (error) {
      throw error;
    }
  }

  @Post('')
  async create(
    @Body(new ZodValidationPipe(userCreateSchema)) createUserDto: userType,
  ): Promise<User> {
    try {
      const userExists = await this.userService.findByUsername(
        createUserDto.data.username,
      );

      if (userExists) {
        throw new HttpException(
          'User with given username already exists',
          HttpStatus.CONFLICT,
        );
      }

      return await this.userService.create(createUserDto);
    } catch (error) {
      throw error;
    }
  }

  @Post('random')
  async createRandomUser() {
    try {
      return await this.userService.createRandomUser();
    } catch (error) {
      throw error;
    }
  }

  @Get(':id')
  async findById(
    @Param('id', new ZodValidationPipe(idSchema)) id: string,
  ): Promise<User> {
    try {
      const result = await this.userService.findById(id);

      if (!result) {
        throw new HttpException(
          'User with given id does not exist in the database',
          HttpStatus.NOT_FOUND,
        );
      }

      return result;
    } catch (error) {
      throw error;
    }
  }

  @Patch(':id')
  async update(
    @Param('id', new ZodValidationPipe(idSchema)) id: string,
    @Body(new ZodValidationPipe(userUpdateSchema))
    body: DeepPartial<userType>,
  ): Promise<User> {
    try {
      const userExists = await this.userService.findById(id);

      if (!userExists) {
        throw new HttpException(
          'User with given username does not exist',
          HttpStatus.NOT_FOUND,
        );
      }
      return await this.userService.update(id, body);
    } catch (error) {
      throw error;
    }
  }

  @Delete(':id')
  async delete(
    @Param('id', new ZodValidationPipe(idSchema)) id: string,
  ): Promise<User> {
    try {
      const userExists = await this.userService.findById(id);

      if (!userExists) {
        throw new HttpException(
          'User with given username does not exist',
          HttpStatus.NOT_FOUND,
        );
      }
      return await this.userService.delete(id);
    } catch (error) {
      throw error;
    }
  }
}
