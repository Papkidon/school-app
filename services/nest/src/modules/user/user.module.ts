import { Module } from '@nestjs/common';
import { UserController } from './user.controller';
import { UserService } from './user.service';
import { UserRepository } from './user.repository';
import { LoggerModule } from 'nestjs-pino';
import { PrismaModule, PrismaService } from 'nestjs-prisma';

@Module({
  imports: [
    LoggerModule.forRoot({
      pinoHttp: {
        transport: {
          target: 'pino-pretty',
          options: {
            singleLine: true,
          },
        },
      },
    }),
    PrismaModule.forRoot(),
  ],
  controllers: [UserController],
  providers: [UserService, UserRepository, PrismaService],
})
export class UserModule {}
