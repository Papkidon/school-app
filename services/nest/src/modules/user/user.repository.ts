import { Injectable } from '@nestjs/common';
import { User } from '@prisma/client';
import { PrismaService } from 'nestjs-prisma';
import { userType } from '../../models/user/user.types';
import { DeepPartial } from '../../models/utility.types';

@Injectable()
export class UserRepository {
  constructor(private prisma: PrismaService) {}
  async findAll(): Promise<User[]> {
    return await this.prisma.user.findMany();
  }

  async findById(id: string): Promise<User> {
    return await this.prisma.user.findUnique({ where: { id } });
  }

  async findByUsername(username: string): Promise<User> {
    return await this.prisma.user.findUnique({ where: { username } });
  }

  async create(body: any): Promise<User> {
    return await this.prisma.user.create(body);
  }

  async createRandomUser(username: string, password: string): Promise<User> {
    return await this.prisma.user.create({
      data: {
        username,
        password,
      },
    });
  }

  async delete(id: string): Promise<User> {
    return await this.prisma.user.delete({ where: { id } });
  }

  async update(id: string, body: DeepPartial<userType>): Promise<User> {
    return await this.prisma.user.update({
      where: { id },
      data: { ...body.data },
    });
  }
}
