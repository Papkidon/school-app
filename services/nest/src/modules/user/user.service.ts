import { Injectable } from '@nestjs/common';
import { UserRepository } from './user.repository';
import { userType } from '../../models/user/user.types';
import { DeepPartial } from '../../models/utility.types';
import { User } from '@prisma/client';
import BuildPerson from '../../utils/BuildPerson/BuildPerson';

@Injectable()
export class UserService {
  constructor(private userRepository: UserRepository) {}
  async findAll(): Promise<User[]> {
    return await this.userRepository.findAll();
  }

  async findById(id: string): Promise<User> {
    return await this.userRepository.findById(id);
  }

  async create(body: any): Promise<User> {
    return await this.userRepository.create(body);
  }

  async createRandomUser(): Promise<User> {
    const randomUser = new BuildPerson().randomize().build();
    return await this.userRepository.createRandomUser(...randomUser);
  }

  async delete(id: string): Promise<User> {
    return await this.userRepository.delete(id);
  }

  async update(id: string, body: DeepPartial<userType>): Promise<User> {
    return await this.userRepository.update(id, body);
  }
}
