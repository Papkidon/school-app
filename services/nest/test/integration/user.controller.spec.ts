import { INestApplication } from '@nestjs/common';
import { Test } from '@nestjs/testing';
import * as request from 'supertest';
import { UserModule } from '../../src/modules/user/user.module';
import { UserRepository } from '../../src/modules/user/user.repository';
import { dummyPersonCreate } from './dummies/dummy.data';

describe('[UserController] Integration tests', () => {
  let app: INestApplication;
  const userRepository = {
    findAll: () => ['dummyFoundUser'],
    findById: () => 'dummyFoundUserById',
    findByUsername: () => 'dummyFoundUserByUsername',
    create: () => 'dummyCreatedUser',
    createRandomUser: () => 'dummyCreatedRandomUser',
    update: () => 'dummyUpdatedUser',
    delete: () => 'dummyDeletedUser',
  };

  beforeAll(async () => {
    const moduleRef = await Test.createTestingModule({
      imports: [UserModule],
    })
      .overrideProvider(UserRepository)
      .useValue(userRepository)
      .compile();

    app = moduleRef.createNestApplication();
    await app.init();
  });

  it(`Get all users - should pass when at least one user exists in the db`, async () => {
    const response = await request(app.getHttpServer()).get('/api/v1/user/');

    expect(response.status).toEqual(200);
    expect(response.body).toMatchSnapshot();
  });

  it(`Get all user - should fail when no user data exists in the db`, async () => {
    jest.spyOn(userRepository, 'findAll').mockImplementationOnce(() => []);

    const response = await request(app.getHttpServer()).get('/api/v1/user/');

    expect(response.status).toEqual(404);
    expect(response.body).toMatchSnapshot();
  });

  it(`Get user by id - should pass when user with given id exists in the db`, async () => {
    const response = await request(app.getHttpServer()).get(
      '/api/v1/user/e4067cdc-dc91-41a6-bbb3-650fdd352dd4',
    );

    expect(response.status).toEqual(200);
    expect(response.body).toMatchSnapshot();
  });

  it(`Get user by id - should fail when user with given id does not exists in the db`, async () => {
    jest.spyOn(userRepository, 'findById').mockImplementationOnce(() => '');

    const response = await request(app.getHttpServer()).get(
      '/api/v1/user/e4067cdc-dc91-41a6-bbb3-650fdd352dd4',
    );

    expect(response.status).toEqual(404);
    expect(response.body).toMatchSnapshot();
  });

  it('Get user by id - should fail when id parameter is not UUID', async () => {
    const response = await request(app.getHttpServer()).get('/api/v1/user/123');

    expect(response.status).toEqual(422);
    expect(response.body).toMatchSnapshot();
  });

  it(`Create user - should pass when request body is correct`, async () => {
    jest
      .spyOn(userRepository, 'findByUsername')
      .mockImplementationOnce(() => '');

    const response = await request(app.getHttpServer())
      .post('/api/v1/user/')
      .send(dummyPersonCreate);

    expect(response.status).toEqual(201);
    expect(response.body).toMatchSnapshot();
  });

  it(`Create user - should fail when user with given username already exists`, async () => {
    const response = await request(app.getHttpServer())
      .post('/api/v1/user/')
      .send(dummyPersonCreate);

    expect(response.status).toEqual(409);
    expect(response.body).toMatchSnapshot();
  });

  it('Create user - should fail when request body is incorrect', async () => {
    const response = await request(app.getHttpServer())
      .post('/api/v1/user/')
      .send({ wrong: 'body' });

    expect(response.status).toEqual(422);
    expect(response.body).toMatchSnapshot();
  });

  it('Create random user - should pass', async () => {
    const response = await request(app.getHttpServer())
      .post('/api/v1/user/random/')
      .send();

    expect(response.status).toEqual(201);
    expect(response.body).toMatchSnapshot();
  });

  it('Create random user - should fail due to an unpredicted error', async () => {
    jest
      .spyOn(userRepository, 'createRandomUser')
      .mockImplementationOnce(() => {
        throw new Error('Something went wrong');
      });

    const response = await request(app.getHttpServer())
      .post('/api/v1/user/random/')
      .send();

    expect(response.status).toEqual(500);
    expect(response.body).toMatchSnapshot();
  });

  it(`Update user - should pass when user with given id exists in the db`, async () => {
    jest.spyOn(userRepository, 'findById').mockImplementationOnce(() => 'xD');
    const response = await request(app.getHttpServer()).patch(
      '/api/v1/user/e4067cdc-dc91-41a6-bbb3-650fdd352dd4',
    ); // It passes without the body, because every property is optional

    expect(response.status).toEqual(200);
    expect(response.body).toMatchSnapshot();
  });

  it(`Update user - should fail when user with given id does not exists in the db`, async () => {
    jest.spyOn(userRepository, 'findById').mockImplementationOnce(() => '');

    const response = await request(app.getHttpServer()).patch(
      '/api/v1/user/e4067cdc-dc91-41a6-bbb3-650fdd352dd4',
    ); // It passes without the body, because every property is optional

    expect(response.status).toEqual(404);
    expect(response.body).toMatchSnapshot();
  });

  it(`Update user - should fail when id parameter is not UUID`, async () => {
    const response = await request(app.getHttpServer()).patch(
      '/api/v1/user/123',
    ); // It passes without the body, because every property is optional

    expect(response.status).toEqual(422);
    expect(response.body).toMatchSnapshot();
  });

  it(`Delete user - should pass when user with given id exists in the db`, async () => {
    const response = await request(app.getHttpServer()).delete(
      '/api/v1/user/e4067cdc-dc91-41a6-bbb3-650fdd352dd4',
    );

    expect(response.status).toEqual(200);
    expect(response.body).toMatchSnapshot();
  });

  it(`Delete user - should fail when user with given id does not exist in the db`, async () => {
    jest.spyOn(userRepository, 'findById').mockImplementationOnce(() => '');

    const response = await request(app.getHttpServer()).delete(
      '/api/v1/user/e4067cdc-dc91-41a6-bbb3-650fdd352dd4',
    );

    expect(response.status).toEqual(404);
    expect(response.body).toMatchSnapshot();
  });

  it(`Delete user - should fail when id parameter is not UUID`, async () => {
    const response = await request(app.getHttpServer()).delete(
      '/api/v1/user/123',
    );

    expect(response.status).toEqual(422);
    expect(response.body).toMatchSnapshot();
  });

  afterAll(async () => {
    await app.close();
  });
});
