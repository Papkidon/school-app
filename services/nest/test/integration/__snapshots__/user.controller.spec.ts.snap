// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[UserController] Integration tests Create random user - should fail due to an unpredicted error 1`] = `
{
  "message": "Internal server error",
  "statusCode": 500,
}
`;

exports[`[UserController] Integration tests Create random user - should pass 1`] = `{}`;

exports[`[UserController] Integration tests Create user - should fail when request body is incorrect 1`] = `
[
  {
    "code": "invalid_type",
    "expected": "object",
    "message": "Required",
    "path": [
      "data",
    ],
    "received": "undefined",
  },
  {
    "code": "unrecognized_keys",
    "keys": [
      "wrong",
    ],
    "message": "Unrecognized key(s) in object: 'wrong'",
    "path": [],
  },
]
`;

exports[`[UserController] Integration tests Create user - should fail when user with given username already exists 1`] = `
{
  "message": "User with given username already exists",
  "statusCode": 409,
}
`;

exports[`[UserController] Integration tests Create user - should pass when request body is correct 1`] = `{}`;

exports[`[UserController] Integration tests Delete user - should fail when id parameter is not UUID 1`] = `
[
  {
    "code": "invalid_string",
    "message": "Invalid uuid",
    "path": [],
    "validation": "uuid",
  },
]
`;

exports[`[UserController] Integration tests Delete user - should fail when user with given id does not exist in the db 1`] = `
{
  "message": "User with given username does not exist",
  "statusCode": 404,
}
`;

exports[`[UserController] Integration tests Delete user - should pass when user with given id exists in the db 1`] = `{}`;

exports[`[UserController] Integration tests Get all user - should fail when no user data exists in the db 1`] = `
{
  "message": "No user data exists in the database",
  "statusCode": 404,
}
`;

exports[`[UserController] Integration tests Get all users - should pass when at least one user exists in the db 1`] = `
[
  "dummyFoundUser",
]
`;

exports[`[UserController] Integration tests Get user by id - should fail when id parameter is not UUID 1`] = `
[
  {
    "code": "invalid_string",
    "message": "Invalid uuid",
    "path": [],
    "validation": "uuid",
  },
]
`;

exports[`[UserController] Integration tests Get user by id - should fail when user with given id does not exists in the db 1`] = `
{
  "message": "User with given id does not exist in the database",
  "statusCode": 404,
}
`;

exports[`[UserController] Integration tests Get user by id - should pass when user with given id exists in the db 1`] = `{}`;

exports[`[UserController] Integration tests Update user - should fail when id parameter is not UUID 1`] = `
[
  {
    "code": "invalid_string",
    "message": "Invalid uuid",
    "path": [],
    "validation": "uuid",
  },
]
`;

exports[`[UserController] Integration tests Update user - should fail when user with given id does not exists in the db 1`] = `
{
  "message": "User with given username does not exist",
  "statusCode": 404,
}
`;

exports[`[UserController] Integration tests Update user - should pass when user with given id exists in the db 1`] = `{}`;
